persistence/trees/trees.go

```diff
@@ -247,6 +255,77 @@ func (t *TreeStore) getStateHash() string {
 	return hexHash
 }
 
+////////////////////////////////
+// AtomicStore Implementation //
+////////////////////////////////
+
+// Savepoint generates a new savepoint for the tree store and saves it internally.
+func (t *TreeStore) Savepoint() error {
+	w, err := t.save()
+	if err != nil {
+		return err
+	}
+	t.Prev = w
+	return nil
+}
+
+// Commit applies any changes that have been made to the underlying trees.
+func (t *TreeStore) Commit() error {
+	if err := t.rootTree.tree.Commit(); err != nil {
+		return err
+	}
+	for treeName, stateTree := range t.merkleTrees {
+		if err := stateTree.tree.Commit(); err != nil {
+			return fmt.Errorf("failed to commit %s: %w", treeName, err)
+		}
+	}
+	return nil
+}
+
+// Rollback intentionally can't return an error because at this point we're out of tricks
+// to recover from problems.
+func (t *TreeStore) Rollback() {
+	if t.Prev != nil {
+		t.merkleTrees = t.Prev.MerkleTrees
+		t.rootTree = t.Prev.RootTree
+		return
+	}
+	t.logger.Fatal().Msgf("rollback called without valid savepoint - this is a bug - treeStore shutting down: %+v", t)
+}
+
+// save commits any pending changes to the trees and serializes the treeStore to a WorldState object
+// with new readers and writers.
+func (t *TreeStore) save() (*Worldstate, error) {
+	if err := t.Commit(); err != nil {
+		return nil, err
+	}
+
+	w := &Worldstate{
+		TreeStoreDir: t.treeStoreDir,
+		MerkleTrees:  map[string]*stateTree{},
```

Comment:
```suggestion
		MerkleTrees:  map[string][]byte,
```
My personal opinion here is that we should store the tree name and root hash and then during the load phase we should do the importing. This would save time as we are more likely to call `save` a lot more than `load` so allocating the SMTs in memory each save seems overkill when it is during load that we need them.

During load we can use `w.TreeStoreDir` and the same pattern we use during `setupTrees()` to get each tree's nodestore

Reply:
This approach comes at the expense of complicating `Rollback` logic with the load functionality rather than the simple reassignment that `TreeStore` uses right now. That could have longer term consequences where maybe a node gets itself into a position where it can't rollback but it can't move forward, either. Having less complicated and stronger guarantees at the `Rollback` point is safer, IMO.

I suggest addressing this in #897 because saving a `Worldstate` is slightly different than making a `Savepoint`, and I think that's important to handle correctly. I'll add an `OPTIMIZE` comment though, because I think that approach is worth exploring.